Fetch all PRs and search in the result

weeman1337 opened the pull request:
Fixes not finding some PRs such as https://github.com/matrix-org/matrix-react-sdk/pull/9002

For some reason a query with the head param does not always return a
result. Fetching all PRs return these as well. So we fetch all and then
iterate the results to find the PR we are looking for.

t3chguy reviewed (changes requested):
Needs pagination support otherwise will only work for the latest ~50 PRs iirc

weeman1337 commented:
> Needs pagination support otherwise will only work for the latest ~50 PRs iirc

As far as I understood from the [docs](https://octokit.github.io/rest.js/v18#pagination) `github.paginate()` should fetch all pages.

t3chguy commented:
Oh yes, the lack of syntax highlight made me miss that call somehow.

Do you have a test run of this somewhere?

weeman1337 commented:
Updated the script:
- First attempt as usual (since it finds most of the PRs)
- Then fallback to pagination
  - Order descending by updated for a high probability to find the PR on the first page
  - Only fetch pages until the PR is found

This is my test code:

```
const { Octokit } = require("@octokit/rest");

const github = new Octokit({
    auth: '...'
})

async function test() {
    const [owner, repo] = "matrix-org/matrix-react-sdk".split("/");
    const head = "weeman1337:feat-dm-on-first";
    //const head = "matrix-org:SimonBrandner/fix/spotlight-btns";
    //const head = "matrix-org:does-not-exist";
    //const head = "tchapgouv:estellecomment/remove-add-space-button";

    // first attempt: query API using the head param
    const response = await github.rest.pulls.list({ head, owner, repo });
    const [queryPR] = response.data;
    // check the head as well because GitHub returns the latest PR if it doesn't find the head
    if (queryPR) return queryPR;

    // For some reason GitHub does not always find the PRs querying with the head param:
    // Fetch all PRs and search for the head.
    let pr;
    await github.paginate(
        "GET /repos/{owner}/{repo}/pulls",
        {
            owner,
            repo,
            sort: "updated",
        },
        (response, done) => {
            const paginatedPR = response.data.find(pr => pr.head?.label === head);
            if (paginatedPR) {
                pr = paginatedPR;
                done();
            }
        }
    );

    if (pr) return pr;
}

test().then((pr) => {
    if (pr) {
        console.log(pr.title, pr.id);
    } else {
        console.log("net gefunden");
    }
});
```

t3chguy commented:
> This is my test code:

So you haven't tested a version of this action using the code though to verify it end-to-end? You can specify action versions by git sha to refer to your unmerged code as version `#9d140e35de7346b58f7730dc50cb2d313f4dc6ed`

t3chguy commented on action.yml:
```diff
@@ -51,7 +51,7 @@ runs:
             },
             (response, done) => {
-              const paginatedPR = response.data.find(pr => pr.head?.label === head);
+              const paginatedPR = response.data.find(pr => pr.head && pr.head.label === head);
```
Could we upgrade github-script to the Node 16 compatible one instead?
